core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -701,7 +701,7 @@ private[spark] object Utils extends Logging {
     }
   }
 
-  private var customHostname: Option[String] = None
+  private var customHostname: Option[String] = sys.env.get("SPARK_LOCAL_HOSTNAME")
```

Comment:
Also note that SPARK_LOCAL_HOSTNAME is actually the DNS we want to advertise internally, not externally, so I don't think it's the same as SPARK_PUBLIC_DNS in DBC's case.

For instance, the external DNS is ec2-x-x-x-x.us-west-2.compute.amazonaws.com and the internal one is ip-y-y-y-y.us-west-2.compute.internal.